### PR TITLE
Provide overloads to avoid unnecessary boxing

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 1]
```diff
@@ -304,6 +324,34 @@ public static ImmutableArray<byte> ToImmutable(this MemoryStream stream)
...
+        /// <param name="array">The array to transform</param>
+        /// <param name="predicate">The condition to use for filtering the array content.</param>
+        /// <param name="selector">A transform function to apply to each element that is not filtered out by <paramref name="predicate"/>.</param>
+        /// <returns>If the items's length is 0, this will return an empty immutable array.</returns>
+        public static ImmutableArray<TResult> SelectManyAsArray<TItem, TResult>(this ImmutableArray<TItem> array, Func<TItem, bool> predicate, Func<TItem, ImmutableArray<TResult>> selector)
+        {
+            if (array.Length == 0)
+            {
+                return ImmutableArray<TResult>.Empty;
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:341]:
```suggestion
            if (array.Length == 0)
                return ImmutableArray<TResult>.Empty;
```

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 2]
```diff
@@ -304,6 +324,34 @@ public static ImmutableArray<byte> ToImmutable(this MemoryStream stream)
...
+
+            var builder = ArrayBuilder<TResult>.GetInstance();
+            foreach (var item in array)
+            {
+                if (predicate(item))
+                {
+                    builder.AddRange(selector(item));
+                }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:349]:
```suggestion
                if (predicate(item))
                    builder.AddRange(selector(item));
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 3]
```diff
@@ -288,17 +308,37 @@ public static ImmutableArray<byte> ToImmutable(this MemoryStream stream)
...
+        /// <summary>
+        /// Maps and flattens a subset of immutable array to another immutable array.
+        /// </summary>
+        /// <typeparam name="TItem">Type of the source array items</typeparam>
+        /// <typeparam name="TResult">Type of the transformed array items</typeparam>
+        /// <param name="array">The array to transform</param>
+        /// <param name="predicate">The condition to use for filtering the array content.</param>
+        /// <param name="selector">A transform function to apply to each element that is not filtered out by <paramref name="predicate"/>.</param>
+        /// <returns>If the items's length is 0, this will return an empty immutable array.</returns>
+        public static ImmutableArray<TResult> SelectManyAsArray<TItem, TResult>(this ImmutableArray<TItem> array, Func<TItem, bool> predicate, Func<TItem, ImmutableArray<TResult>> selector)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:332]: 📝 There were two callers of the previous overload prior to this change. After this change, there is one caller of the previous overload, and one caller of the new overload.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 4]
```diff
@@ -276,6 +276,26 @@ public static ImmutableArray<byte> ToImmutable(this MemoryStream stream)
...
 
+        /// <summary>
+        /// Maps and flattens a subset of immutable array to another immutable array.
+        /// </summary>
+        /// <typeparam name="TItem">Type of the source array items</typeparam>
+        /// <typeparam name="TResult">Type of the transformed array items</typeparam>
+        /// <param name="array">The array to transform</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>If the array's length is 0, this will return an empty immutable array.</returns>
+        public static ImmutableArray<TResult> SelectManyAsArray<TItem, TResult>(this ImmutableArray<TItem> array, Func<TItem, ImmutableArray<TResult>> selector)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:287]: 📝 There were 16 callers of the previous overload prior to this change. After this change, there are 3 callers of the previous overload, and 13 callers of the new overload.

**Review by 333fred — APPROVED** [review]

**Review by jjonescz — APPROVED** [review]

**Pull request merged**